hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/prehandle/CompoundSignatureVerificationFuture.java

```diff
@@ -0,0 +1,218 @@
+     * cancel it. If it does not exist, we will let it be.
+     */
+    @Override
+    public boolean cancel(final boolean mayInterruptIfRunning) {
+        // If we're already done, then we can't cancel again
+        if (isDone()) {
+            return false;
+        }
+
+        // Try to cancel each underlying future (I go ahead and try canceling all of them, even if one fails).
+        // If the underlying Future doesn't exist, then we just skip it. Best effort.
+        for (final var future : futures) {
+            future.cancel(mayInterruptIfRunning);
+        }
+
+        // Record that we have had "canceled" called already, so we don't do it again, and so that "done" is right.
+        canceled = true;
+        return true;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public boolean isCancelled() {
+        return canceled;
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * Due to the asynchronous nature of {@link TransactionSignature}, it is possible that at the time this method
+     * is called, one or more {@link TransactionSignature}s do not yet have a {@link Future}. If that is the case,
+     * then we are clearly not done!
+     */
+    @Override
+    public boolean isDone() {
+        if (canceled) {
+            return true;
+        }
+
+        for (final var future : futures) {
+            if (!future.isDone()) {
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * Blocks on each {@link Future}. Due to the asynchronous nature of {@link TransactionSignature}, it is possible
+     * that a {@link Future} has not been assigned to one or more {@link TransactionSignature}s. In that case, we will
+     * wait for the {@link Future} before proceeding.
+     */
+    @Override
+    public SignatureVerification get() throws InterruptedException, ExecutionException {
+        for (final var future : futures) {
+            final var verification = future.get();
+            if (!verification.passed()) {
+                return new SignatureVerificationImpl(key, hollowAccount, false);
+            }
+        }
+
+        return new SignatureVerificationImpl(key, hollowAccount, true);
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * Blocks on each {@link Future}. Due to the asynchronous nature of {@link TransactionSignature}, it is possible
+     * that a {@link Future} has not been assigned to one or more {@link TransactionSignature}s. In that case, we will
+     * wait for the {@link Future} before proceeding (up to the timeout).
+     */
+    @Override
+    public SignatureVerification get(final long timeout, final TimeUnit unit)
+            throws InterruptedException, ExecutionException, TimeoutException {
+        var millisRemaining = unit.toMillis(timeout);
+        int failCount = 0;
+        for (final var future : futures) {
```

Comment:
Might be good to use CompletableFuture.allOf to set a single timeout, then process all of the items (counting success and failure for your failCount entries)
```
// wait for all to complete, or timeout
CompletableFuture.allOf(arrayOfCompletableFutures).orTimeout(timeout, unit).join();
// process each future in the array
```

Reply:
A CompletableFuture is complete when it's value is available, it's not generally completed by code outside the task it represents.  I would claim that any Future that returns a single result value can be considered a CompletableFuture.